include/libpmemobj++/array_iterator.hpp

```diff
@@ -0,0 +1,341 @@
+		return *static_cast<Iterator *>(this);
+	}
+
+	/**
+	* Postfix increment operator
+	*/
+	Iterator operator++(int)
+	{
+		Iterator tmp(*static_cast<Iterator *>(this));
+		static_cast<Iterator *>(this)->change_by(1);
+		return tmp;
+	}
+
+	/**
+	* Prefix decrement operator
+	*/
+	Iterator &operator--()
+	{
+		static_cast<Iterator *>(this)->change_by(-1);
+		return *static_cast<Iterator *>(this);
+	}
+
+	/**
+	* Postfix decrement operator
+	*/
+	Iterator operator--(int)
+	{
+		Iterator tmp(*static_cast<Iterator *>(this));
+		static_cast<Iterator *>(this)->change_by(-1);
+		return tmp;
+	}
+
+	/**
+	 * Addition assignment operator.
+	 */
+	Iterator &
+	operator+=(std::ptrdiff_t n)
+	{
+		static_cast<Iterator *>(this)->change_by(n);
+		return *static_cast<Iterator *>(this);
+	}
+
+	/**
+	* Subtraction assignment operator.
+	*/
+	Iterator &
+	operator-=(std::ptrdiff_t n)
+	{
+		static_cast<Iterator *>(this)->change_by(-n);
+		return *static_cast<Iterator *>(this);
+	}
+
+	/**
+	 * Addition operator
+	 */
+	Iterator
+	operator+(std::ptrdiff_t n)
+	{
+		Iterator tmp(*static_cast<Iterator *>(this));
+		tmp += n;
+		return tmp;
+	}
+
+	/**
+	 * Subtraction operator overload for integral type
+	 */
+	Iterator
+	operator-(std::ptrdiff_t n)
+	{
+		Iterator tmp(*static_cast<Iterator *>(this));
+		tmp -= n;
+		return tmp;
+	}
+
+	/**
+	 * Subtraction operator overload Iterator type
+	 */
+	friend std::ptrdiff_t
+	operator-(const Iterator &lhs, const Iterator &rhs)
+	{
+		return lhs.ptr - rhs.ptr;
+	}
+
+	/**
+	 * Element access operator
+	 */
+	Reference operator[](std::ptrdiff_t n)
+	{
+		return ptr[n];
+	}
+
+	/**
+	 * Non-member equal operator
+	 */
+	constexpr friend bool
+	operator==(const Iterator &lhs, const Iterator &rhs)
+	{
+		return lhs.ptr == rhs.ptr;
+	}
+
+	/**
+	 * Non-member not equal operator
+	 */
+	constexpr friend bool
+	operator!=(const Iterator &lhs, const Iterator &rhs)
+	{
+		return !(lhs == rhs);
+	}
+
+	/**
+     * Non-member less than operator
+     */
+	constexpr friend bool
+	operator<(const Iterator &lhs, const Iterator &rhs)
+	{
+		return lhs.ptr < rhs.ptr;
+	}
+
+	/**
+	 * Non-member greater than operator
+	 */
+	constexpr friend bool
+	operator>(const Iterator &lhs, const Iterator &rhs)
+	{
+		return lhs.ptr > rhs.ptr;
+	}
+
+	/**
+	 * Non-member less or equal operator
+	 */
+	constexpr friend bool
+	operator<=(const Iterator &lhs, const Iterator &rhs)
+	{
+		return !(lhs > rhs);
+	}
+
+	/**
+	 * Non-member greater or equal operator
+	 */
+	constexpr friend bool
+	operator>=(const Iterator &lhs, const Iterator &rhs)
+	{
+		return !(lhs < rhs);
+	}
+
+protected:
+	/*
+	 * Function for changing underlying pointer.
+	 * Non-const iterator may override it and snapshot data if necessary.
+	 */
+	void
+	change_by(std::ptrdiff_t n)
+	{
+		ptr += n;
+	}
+
+	Pointer ptr;
+};
+
+template <typename T>
+struct const_array_iterator;
+
+template <typename T>
+struct array_iterator : public base_iterator<array_iterator<T>, T &, T *> {
```

Comment:
I thought the entire point of using the snapshot closure is to generalize this interface so that it does not assume that the objects are laid out contiguously.

Why is this `array_iterator`? Will we have a `vector_iterator` for vector? If we must do so, I'd prefer to put this iterator in the `pmem::obj:array::iterator` class.